src/subsystem/can/commands.cpp

```diff
@@ -3,72 +3,307 @@
+    
+    while ((read_size <= 0 || frame.can_id != expected_can_id) && delta_time.count() <= MAX_READ_TIME) {
+        read_size = read(can_socket, &frame, sizeof(struct can_frame));
+        delta_time = std::chrono::steady_clock::now() - start_time;
+    }
+
+    //Check if invalid read
+    if (read_size <= 0 || frame.can_id != expected_can_id) {
+        status = CAN_Status::FAILED_READ;
+        return 0;
+    }
+
+    //Put data into unsigned integer
+    for (int i = 0; i < 4; i++) {
+        return_value |= ((unsigned int)(frame.data[i]) << (8 * i));
+    }
+    
+#endif
+    //Successfully read
+    status = CAN_Status::SUCCESS;
+    return return_value;
+}
+
+//receive a CAN message
+long can_receive_long(Node device, Command command) {
+    //Define return value
+    unsigned int return_value = 0;
+
+    //Set status to unset
+    status = CAN_Status::UNSET;
+
+#ifdef ONBOARD_CAN_BUS
+    //Socket is not open, nothing can read
+    if (!socket_open) { return 1; }
+    
+    //Create the frame
+    can_frame frame;
+    frame = get_can_frame(0, device, command, 0, 0);
+    int expected_can_id = command | (device << 5);
+    frame.can_id = expected_can_id;
+
+    //Set the filter
+    struct can_filter rfilter[1];
+	rfilter[0].can_id = expected_can_id;
+	rfilter[0].can_mask = expected_can_id;
+	setsockopt(can_socket, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+
+    //Read the frame
+    int read_size = 0;
+    while (read_size <= 0 && frame.can_id != expected_can_id) {
+        read_size = read(can_socket, &frame, sizeof(struct can_frame));
+    }
+
+    //Check if invalid read
+    if (read_size <= 0) {
+        status = CAN_Status::FAILED_READ;
+        return 0;
+    }
+
+    //Put data into long
+    for (int i = 0; i < read_size; i++) {
+        return_value |= ((long)(frame.data[i]) << (8 * i));
+    }
+    
+#endif
+    //Successfully read
+    status = CAN_Status::SUCCESS;
+    return return_value;
+}
+
+//Recieve a CAN frame and check it's ID (for now, this only works for odrives)
+bool can_check_hearbeat(Node device) {
+    //Set status to unset
+    status = CAN_Status::UNSET;
+    bool return_value = true;
+
+#ifdef ONBOARD_CAN_BUS
+    //Socket is not open, nothing can read
+    if (!socket_open) { return 1; }
+    
+    //Create a frame
+    can_frame frame;
+
+    //Read current frame
+    int expected_can_id = (Command::DRIVE_HEARTBEAT_MESSAGE) || (device << 5);
+    int read_size = read(can_socket, &frame, sizeof(struct can_frame));
+
+    //Check if invalid read
+    if (read_size < 0) {
+        status = CAN_Status::FAILED_READ;
+        return 0;
+    }
+    
+    //Set return value
+    return_value = (frame.can_id == expected_can_id);
+
+#endif
+    //Successfully read
+    status = CAN_Status::SUCCESS;
+    return return_value;
+}
+
+//Receive all the vital information from the control teensy
+float* get_control_information() {
+    //return information
+    float* ret = new float[10];
```

Comment:
Instead of dynamically allocating an array, we should define a struct that contains this info with named fields. The caller can supply a reference to this struct for this method to populate